File: test/_template.py
import sys

sys.path.insert(0, "{root}/py")
import slon

with open('{sources_path}', encoding='utf8') as fp:
    sources = fp.read()

delim = '{delim}'
res = []

HOOKS = {
    'upper': lambda x: x.upper(),
    'add5': lambda x: x + 5,
    'addX': lambda x: x + 'x',
}


def serialize(x):
    if x is None:
        return 'null'
    if isinstance(x, bool):
        return 'true' if x else 'false'
    if isinstance(x, (int, float)):
        return str(x)
    if isinstance(x, str):
        s = ''
        for c in x:
            n = ord(c)
            if 32 <= n <= 127:
                s += c
            else:
                s += r'\u{%x}' % n
        return '"' + s + '"'
    if isinstance(x, list):
        return '[' + ', '.join(serialize(v) for v in x) + ']'
    if isinstance(x, dict):
        return '{' + ', '.join(serialize(k) + ': ' + serialize(v) for k, v in x.items()) + '}'
    raise ValueError('serialization error')


for src in sources.split(delim):
    try:
        r = slon.loads(src, hooks=HOOKS)
        r = serialize(r)
    except slon.SlonError as err:
        r = 'ERROR:%s:%s:%s:%s:%s' % (err.code, err.row, err.col, err.start_row, err.start_col)
    except Exception as err:
        r = 'ERROR:%s' % repr(err)
    res.append(r)

with open('{results_path}', 'wt', encoding='utf8') as fp:
    fp.write(delim.join(res))

File: test/_template.js
let fs = require('fs');
let slon = require('{root}/js/slon')

let sources = fs.readFileSync('{sources_path}', {encoding: 'utf8'})

let delim = '{delim}'
let res = []

HOOKS = {
    'upper': x => x.toUpperCase(),
    'add5': x => x + 5,
    'addX': x => x + 'x',
}


function serialize(x) {
    if (x === null)
        return 'null'
    if (typeof x === 'boolean')
        return x ? 'true' : 'false'
    if (typeof x === 'number')
        return x.toString()
    if (typeof x === 'string') {
        let s = ''
        for (let c of [...x]) {
            let n = c.codePointAt(0)
            if (32 <= n && n <= 127)
                s += c
            else
                s += '\\u{' + n.toString(16) + '}'
        }
        return '"' + s + '"'
    }
    if (Array.isArray(x))
        return '[' + x.map(serialize).join(', ') + ']'
    if (typeof x === 'object')
        return '{' + [...Object.entries(x)].map(kv => serialize(kv[0]) + ': ' + serialize(kv[1])).join(', ') + '}'
    throw new Error('serialization error')
}


for (let src of sources.split(delim)) {
    let r
    try {
        r = slon.parse(src, {hooks: HOOKS})
        r = serialize(r)
    } catch (err) {
        if (err instanceof slon.SlonError)
            r = `ERROR:${err.code}:${err.row}:${err.col}:${err.startRow}:${err.startCol}`
        else
            r = 'ERROR:' + err.toString()
    }
    res.push(r)
}

fs.writeFileSync('{results_path}', res.join(delim), {encoding: 'utf8'})
